File: src/android/CDVMobilePayAppSwitch.java
package com.github.emilbayes.CDVMobilePayAppSwitch;

import android.content.Intent;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.apache.cordova.PluginResult.Status;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.math.BigDecimal;
import java.security.SecureRandom;
import java.util.Random;

import dk.danskebank.mobilepay.sdk.Country;
import dk.danskebank.mobilepay.sdk.MobilePay;
import dk.danskebank.mobilepay.sdk.ResultCallback;
import dk.danskebank.mobilepay.sdk.model.FailureResult;
import dk.danskebank.mobilepay.sdk.model.Payment;
import dk.danskebank.mobilepay.sdk.model.SuccessResult;


public class CDVMobilePayAppSwitch extends CordovaPlugin {
    private CallbackContext _listenerCallback;
    private final Object _listenerCallbackLock = new Object();
    private CallbackContext _inflightPaymentCallback;
    private String _inflightOrderId;
    private int MOBILEPAY_PAYMENT_REQUEST_CODE = 0;

    private Random srand;

    @Override
    public void initialize(final CordovaInterface cordova, final CordovaWebView webView) {
        super.initialize(cordova, webView);

        this.srand = new SecureRandom();
    }

    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("isMobilePayInstalled")) return isMobilePayInstalled(args, callbackContext);
        if (action.equals("attachListener")) return attachListener(args, callbackContext);
        if (action.equals("setupWithMerchantId")) return setupWithMerchantId(args, callbackContext);
        if (action.equals("beginMobilePaymentWithPayment")) return beginMobilePaymentWithPayment(args, callbackContext);

        callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.INVALID_ACTION));
        return false;
    }

    protected boolean isMobilePayInstalled(JSONArray args, CallbackContext callbackContext) throws JSONException {
        Country country = Country.DENMARK; // TODO make multi country
        boolean isInstalled = MobilePay.getInstance().isMobilePayInstalled(this.cordova.getActivity().getApplicationContext(), country);
        callbackContext.sendPluginResult(new PluginResult(Status.OK, isInstalled));

        return true;
    }

    protected boolean attachListener(JSONArray args, CallbackContext callbackContext) throws JSONException {
        synchronized(_listenerCallbackLock) {
            _listenerCallback = callbackContext;
        }
        notifyListenerOfProp("isAppSwitchInProgress", MobilePay.getInstance().hasActivePayment());
        return true;
    }

    protected void notifyListenerOfProp (String prop, Object value) {
        synchronized(_listenerCallbackLock) {
            try {
                PluginResult result = new PluginResult(Status.OK, new JSONArray(new Object[]{prop, value}));
                result.setKeepCallback(true);
                _listenerCallback.sendPluginResult(result);
            } catch (JSONException ex) {
                PluginResult result = new PluginResult(Status.ERROR, "Unable to serialize prop notification");
                result.setKeepCallback(true);
                _listenerCallback.sendPluginResult(result);
            }
        }
    }

    protected boolean setupWithMerchantId(JSONArray args, CallbackContext callbackContext) throws JSONException {
        String merchantId = args.getString(0);

        // Unused by Android
        // String merchantUrlScheme = args.getString(1)

        JSONObject options = args.getJSONObject(2);
        if (options == null) options = new JSONObject(); // Just so we dont have to handle this case

        Country country = Country.DENMARK;

        MobilePay.getInstance().init(merchantId, country);

        int returnSeconds = options.optInt("returnSeconds", 5);
        if (returnSeconds < 0) {
            callbackContext.error("returnSeconds must not be negative");
            return false;
        }
        if (returnSeconds > 9) {
            callbackContext.error("returnSeconds must not be greater than 9s");
            return false;
        }

        int timeoutSeconds = options.optInt("timeoutSeconds", 0);
        if (timeoutSeconds < 0) {
            callbackContext.error("timeoutSeconds must not be negative");
            return false;
        }
        if (timeoutSeconds > 1200) {
            callbackContext.error("timeoutSeconds must not be greater than 1200s");
            return false;
        }

        MobilePay.getInstance().setReturnSeconds(returnSeconds);
        MobilePay.getInstance().setTimeoutSeconds(timeoutSeconds);

        callbackContext.sendPluginResult(new PluginResult(Status.OK, true));

        return true;
    }

    protected boolean beginMobilePaymentWithPayment(JSONArray args, CallbackContext callbackContext) throws JSONException {
        MOBILEPAY_PAYMENT_REQUEST_CODE = this.srand.nextInt(65535) + 1;

        String orderId = args.getString(0);
        if (orderId.getBytes().length < 4) {
            callbackContext.error("Too short orderId (must be at least 4 chars)");
            return false;
        }

        if (orderId.getBytes().length > 50) {
            callbackContext.error("Too long orderId (must be at most 50 chars)");
            return false;
        }

        double productPrice = args.getDouble(1);
        if (productPrice < 0) {
            callbackContext.error("productPrice must be greater than zero");
            return false;
        }

        // MobilePay is present on the system. Create a Payment object.
        Payment payment = new Payment();
        payment.setProductPrice(new BigDecimal(productPrice));
        payment.setOrderId(orderId);

        // Create a payment Intent using the Payment object from above.
        Intent paymentIntent = MobilePay.getInstance().createPaymentIntent(payment);

        // We now jump to MobilePay to complete the transaction. Start MobilePay and wait for the result using an unique result code of your choice.
        cordova.setActivityResultCallback (this);
        cordova.startActivityForResult(this, paymentIntent, MOBILEPAY_PAYMENT_REQUEST_CODE);
        notifyListenerOfProp("isAppSwitchInProgress", MobilePay.getInstance().hasActivePayment());
        _inflightOrderId = orderId; // need to fake this one on Android
        _inflightPaymentCallback = callbackContext;

        return true;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == MOBILEPAY_PAYMENT_REQUEST_CODE) {
            MOBILEPAY_PAYMENT_REQUEST_CODE = 0;

            // The request code matches our MobilePay Intent
            MobilePay.getInstance().handleResult(resultCode, data, new ResultCallback() {
                @Override
                public void onSuccess(SuccessResult result) {
                    notifyListenerOfProp("isAppSwitchInProgress", MobilePay.getInstance().hasActivePayment());
                    try {
                        JSONObject map = new JSONObject();
                        map.put("success", true);
                        map.put("cancelled", false);
                        map.put("orderId", result.getOrderId());
                        map.put("productPrice", 10);
                        map.put("amountWithdrawnFromCard", result.getAmountWithdrawnFromCard());
                        map.put("transactionId", result.getTransactionId());
                        map.put("signature", result.getSignature());

                        _inflightPaymentCallback.success(map);
                    } catch (JSONException ex) {
                        _inflightPaymentCallback.error("Unable to serialise success result");
                    } finally {
                        _inflightPaymentCallback = null;
                        _inflightOrderId = null;
                    }
                }
                @Override
                public void onFailure(FailureResult result) {
                    notifyListenerOfProp("isAppSwitchInProgress", MobilePay.getInstance().hasActivePayment());
                    try {
                        JSONObject map = new JSONObject();
                        map.put("success", false);
                        map.put("cancelled", false);
                        map.put("orderId", result.getOrderId());
                        map.put("errorCode", result.getErrorCode());
                        map.put("errorMessage", result.getErrorMessage());

                        _inflightPaymentCallback.success(map);
                    } catch (JSONException ex) {
                        _inflightPaymentCallback.error("Unable to serialise error result");
                    } finally {
                        _inflightPaymentCallback = null;
                        _inflightOrderId = null;
                    }
                }
                @Override
                public void onCancel() {
                    notifyListenerOfProp("isAppSwitchInProgress", MobilePay.getInstance().hasActivePayment());
                    try {
                        JSONObject map = new JSONObject();
                        map.put("success", false);
                        map.put("cancelled", true);
                        map.put("orderId", _inflightOrderId);

                        _inflightPaymentCallback.success(map);
                    } catch (JSONException ex) {
                        _inflightPaymentCallback.error("Unable to serialise cancel result");
                    } finally {
                        _inflightPaymentCallback = null;
                        _inflightOrderId = null;
                    }
                }
            });
        }
    }
}

File: src/ios/CDVMobilePayAppSwitch.m
#import <Cordova/CDV.h>
#import <MobilePayManager.h>
#import "CDVMobilePayAppSwitch.h"

@interface CDVMobilePayAppSwitch () {}
@end

@implementation CDVMobilePayAppSwitch

- (void)attachListener:(CDVInvokedUrlCommand*)command
{
    listenerCallback = command;
    [self notifyListenerOfProp:@"isAppSwitchInProgress" value:@([[MobilePayManager sharedInstance] isAppSwitchInProgress])];
}

- (void)notifyListenerOfProp:(NSString*)prop
                       value:(id)value
{
    dispatch_async(dispatch_get_main_queue(), ^{
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:@{@"prop": prop, @"value": value}];
        [pluginResult setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:listenerCallback.callbackId];
    });
}

- (void)isMobilePayInstalled:(CDVInvokedUrlCommand*)command
{
    MobilePayCountry country = MobilePayCountry_Denmark; // TODO make multi country

    BOOL isInstalled = [[MobilePayManager sharedInstance] isMobilePayInstalled:country];
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:isInstalled];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setupWithMerchantId:(CDVInvokedUrlCommand*)command
{

    NSString* merchantId = [command argumentAtIndex:0];
    if ([self assert:(merchantId != nil) errorMessage:@"Missing merchantId" command:command]) return;

    NSString* merchantUrlScheme =  [command argumentAtIndex:1];
    if ([self assert:(merchantUrlScheme != nil) errorMessage:@"Missing merchantUrlScheme" command:command]) return;

    NSDictionary* options = [command argumentAtIndex:2 withDefault:nil andClass:[NSDictionary class]];

    // These ranges come form the disassembled .jar file
    NSNumber* maybeReturnSeconds = [options objectForKey:@"returnSeconds"];
    int returnSeconds = maybeReturnSeconds != nil ? [maybeReturnSeconds intValue] : 5;
    if ([self assert:(returnSeconds >= 0) errorMessage:@"returnSeconds must not be negative" command:command]) return;
    if ([self assert:(returnSeconds <= 9) errorMessage:@"returnSeconds must not be greater than 9s" command:command]) return;

    NSNumber* maybeTimeoutSeconds = [options objectForKey:@"timeoutSeconds"];
    int timeoutSeconds = maybeTimeoutSeconds != nil ? [maybeTimeoutSeconds intValue] : 0;
    if ([self assert:(timeoutSeconds >= 0) errorMessage:@"timeoutSeconds must not be negative" command:command]) return;
    if ([self assert:(timeoutSeconds <= 1200) errorMessage:@"timeoutSeconds must not be greater than 1200s" command:command]) return;

    MobilePayCountry country = MobilePayCountry_Denmark; // TODO make multi country

    [[MobilePayManager sharedInstance]
     setupWithMerchantId:merchantId
     merchantUrlScheme:merchantUrlScheme
     timeoutSeconds:timeoutSeconds
     captureType:MobilePayCaptureType_Reserve // TODO make option
     country:country];

    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}


- (void)beginMobilePaymentWithPayment:(CDVInvokedUrlCommand*)command
{
    if ([self assert:([[MobilePayManager sharedInstance] isAppSwitchInProgress] == NO) errorMessage:@"AppSwitch already in progress" command:command]) return;

    NSString* orderId = [command argumentAtIndex:0];
    if ([self assert:(orderId != nil) errorMessage:@"Missing orderId" command:command]) return;
    // As per https://github.com/MobilePayDev/MobilePay-AppSwitch-SDK/wiki/Parameter-specification
    if ([self assert:([orderId length] >= 4) errorMessage:@"Too short orderId (must be at least 4 chars)" command:command]) return;
    if ([self assert:([orderId length] <= 50) errorMessage:@"Too long orderId (must be at most 50 chars)" command:command]) return;

    NSDecimalNumber* productPrice = [command argumentAtIndex:1];
    if ([self assert:(productPrice != nil) errorMessage:@"Missing productPrice" command:command]) return;
    if ([self assert:(productPrice > 0) errorMessage:@"productPrice must be greater than zero" command:command]) return;

    MobilePayPayment* payment = [[MobilePayPayment alloc] initWithOrderId:orderId productPrice:productPrice];

    inflightPaymentCallbackId = command.callbackId;
    inflightOrderId = orderId;

    [[MobilePayManager sharedInstance] beginMobilePaymentWithPayment:payment error:^(MobilePayErrorPayment * _Nonnull error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self notifyListenerOfProp:@"isAppSwitchInProgress" value:@([[MobilePayManager sharedInstance] isAppSwitchInProgress])];
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary:@{
              @"orderId": inflightOrderId,
              @"errorCode": [NSNumber numberWithInteger:error.error.code],
              @"errorMessage": [error.error.userInfo valueForKey:NSLocalizedFailureReasonErrorKey],
              @"success": @NO,
              @"cancelled": @NO
            }];

            [self.commandDelegate sendPluginResult:pluginResult callbackId:inflightPaymentCallbackId];
            inflightPaymentCallbackId = nil;
            inflightOrderId = nil;
        });
    }];

    [self notifyListenerOfProp:@"isAppSwitchInProgress" value:@([[MobilePayManager sharedInstance] isAppSwitchInProgress])];
}

- (void)handleOpenURL:(NSNotification*)notification {
    NSURL* url = [notification object];

    if (![url isKindOfClass: [NSURL class]]) return;

    [self handleMobilePayPaymentWithUrl:url];
}

- (void)handleMobilePayPaymentWithUrl:(NSURL*)url
{
    [[MobilePayManager sharedInstance]handleMobilePayPaymentWithUrl:url success:^(MobilePaySuccessfulPayment * _Nullable payment) {
        [self notifyListenerOfProp:@"isAppSwitchInProgress" value:@([[MobilePayManager sharedInstance] isAppSwitchInProgress])];
        NSDictionary* result = @{@"orderId": payment.orderId,
                                  @"transactionId": payment.transactionId,
                                  @"signature": payment.signature,
                                  @"amountWithdrawnFromCard": payment.amountWithdrawnFromCard,
                                  @"success": @YES,
                                  @"cancelled": @NO};
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:inflightPaymentCallbackId];
        inflightPaymentCallbackId = nil;
        inflightOrderId = nil;

    } error:^(MobilePayErrorPayment * _Nullable error) {
        [self notifyListenerOfProp:@"isAppSwitchInProgress" value:@([[MobilePayManager sharedInstance] isAppSwitchInProgress])];
        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsDictionary:@{
          @"orderId": inflightOrderId,
          @"errorCode": [NSNumber numberWithInteger:error.error.code],
          @"errorMessage": [error.error.userInfo valueForKey:NSLocalizedFailureReasonErrorKey],
          @"success": @NO,
          @"cancelled": @NO
        }];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:inflightPaymentCallbackId];
        inflightPaymentCallbackId = nil;
        inflightOrderId = nil;
    } cancel:^(MobilePayCancelledPayment * _Nullable payment) {
        [self notifyListenerOfProp:@"isAppSwitchInProgress" value:@([[MobilePayManager sharedInstance] isAppSwitchInProgress])];
        NSDictionary* result = @{@"orderId": payment.orderId,
                                 @"success": @NO,
                                 @"cancelled": @YES};

        CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];

        [self.commandDelegate sendPluginResult:pluginResult callbackId:inflightPaymentCallbackId];
        inflightPaymentCallbackId = nil;
        inflightOrderId = nil;
    }];
}

- (BOOL)assert:(BOOL)predicate
        errorMessage:(NSString*)errorMessage
        command:(CDVInvokedUrlCommand*)command
{
  if (predicate == YES) return NO;

  CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                   messageAsString:errorMessage];

  [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];

  return YES;
}

@end
